urlac47192b66_ggrandes_bouncer_tgz-pJ8-org_javastack_bouncer_ClusterTestJ8/src/org/javastack/bouncer/OutboundAddress.java

```diff
@@ -0,0 +1,214 @@
+					+ fromArrAddress(addrs) + "]");
+		}
+	}
+
+	Socket connect() throws UnknownHostException {
+		return connectFrom(null);
+	}
+
+	Socket connectFrom(final InetAddress stickyAddr) throws UnknownHostException {
+		resolve();
+		if (addrs == null) {
+			return null;
+		}
+		Socket remote = null;
+		// First, try sticky, if any...
+		if ((stickies != null) && (stickyAddr != null)) {
+			final InetAddress dstAddr = stickies.get(stickyAddr);
+			if (dstAddr != null) {
+				for (final InetAddress addr : addrs) {
+					if (addr.equals(dstAddr)) {
+						Log.error(this.getClass().getSimpleName() + " Sticky id=" + stickyAddr + " result="
+								+ addr);
+						remote = connect0(addr);
+						break;
+					}
+				}
+			}
+		}
+		// Else, try another host...
+		if (remote == null) {
+			final int filterFlags = (Options.LB_ORDER | Options.LB_RR | Options.LB_RAND);
+			int begin = 0;
+			InetAddress[] addrs = this.addrs;
+			if (addrs.length > 1) {
+				switch (opts.getFlags(filterFlags)) {
+					case Options.LB_ORDER:
+						break;
+					case Options.LB_RR:
+						begin = roundrobin;
+						break;
+					case Options.LB_RAND:
+						addrs = this.addrs.clone();
+						Collections.shuffle(Arrays.asList(addrs));
+						break;
+				}
+			}
+			// Use local var to avoid synchronized block
+			int rr = begin;
+			do {
+				final InetAddress addr = addrs[rr];
+				remote = connect0(addr);
+				rr = ((rr + 1) % addrs.length);
+				roundrobin = rr;
+				if (remote != null) {
+					break;
+				}
+			} while (roundrobin != begin);
+		}
+		if (remote != null) {
+			try {
+				if ((stickies != null) && (stickyAddr != null)) {
+					final InetAddress currentSticky = stickies.get(stickyAddr);
+					if (!remote.getInetAddress().equals(currentSticky)) {
+						final StickyConfig stickyCfg = opts.getStickyConfig();
+						if (stickyCfg.isReplicated()) {
+							context.stickyLocalUpdateNotify(stickyCfg.clusterId, stickyCfg.replicationId,
+									stickyAddr, remote.getInetAddress());
+						}
+						stickies.put(stickyAddr, remote.getInetAddress());
+					}
+				}
+				context.registerSocket(remote);
+				final Integer pReadTimeout = opts.getInteger(Options.P_READ_TIMEOUT);
+				if (pReadTimeout != null) {
+					remote.setSoTimeout(pReadTimeout);
+				}
+			} catch (SocketException e) {
+				Log.error(this.getClass().getSimpleName() + " Error setting parameters to socket: " + remote);
+			}
+		}
+		return remote;
+	}
+
+	private Socket connect0(final InetAddress dstAddr) {
+		final boolean isSSL = opts.isOption(Options.TUN_SSL | Options.MUX_SSL);
+		Socket sock = null;
+		try {
+			Log.info(getClass().getSimpleName() + " Connecting to " + dstAddr + ":" + port
+					+ (isSSL ? " (SSL)" : ""));
+			if (opts.isOption(Options.MUX_SSL)) {
+				sock = sslFactory.createSSLSocket();
+			} else if (opts.isOption(Options.TUN_SSL) && (sslFactory != null)) {
+				sock = sslFactory.createSSLSocket();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *sock*. The resource is referenced at lines 179, 180, 191, 192, 193, 196, 199, 201, 204, 206, 208. The resource is returned at line 209. There are other execution paths that don&#39;t close the resource or return it, for example, when an else condition at line 206 evaluates to true. Ensure *sock* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).